src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs

```diff
@@ -410,6 +410,89 @@ public void Dispose()
         [ThreadStatic]
         public static int ThreadId;
 
+        private class Foo
+        {
+            public Bar Bar;
+            public Foo(Bar bar)
+            {
+                Bar = bar;
+            }
+        }
+
+        private class Bar
+        {
+            public Bar(ManualResetEvent mre1, ManualResetEvent mre2)
+            {
+                // Making sure ctor gets called only once
+                Assert.True(!mre1.WaitOne(0) && !mre2.WaitOne(0)); 
+
+                // Then use mre2 to signal execution reached this ctor call
+                mre2.Set();
+
+                // Wait until it's OK to leave ctor
+                mre1.WaitOne();
+            }
+        }
+
+        [Fact]
+        public async Task GetRequiredService_ResolvingSameSingletonInTwoThreads_SameServiceReturned()
+        {
+            Bar bar = null;
+            Foo foo = null;
+            IServiceProvider sp = null;
+
+            // Arrange
+            var services = new ServiceCollection();
+
+            services.AddSingleton<Foo>();
+            services.AddSingleton<Bar>(sp => {
+                return new Bar(_mreForThread1, _mreForThread2);
+            });
+
+            sp = services.BuildServiceProvider();
+
+            var t1 = Task.Run(() =>
+            {
+                using var scope1 = sp.CreateScope();
+                foo = scope1.ServiceProvider.GetRequiredService<Foo>();
```

Comment:
Do we need to create a scope here?
```suggestion
                foo = sp.GetRequiredService<Foo>();
```